### PR TITLE
Merge stateless idea with no local storage of public and private keys, support multiple interface per hosts using several groups

### PR DESCRIPTION
**Pull request by fbourqui**
In addition:
- add reload module on update config file

Solve:
#18
#25
#26

#17, is now irrelevant

### CODE REVIEW

**Comment by fbourqui** [comment]
Solve:
#18 
#26

File: tasks/main.yml [hunk 1]
```diff
@@ -28,106 +27,58 @@
...
-- name: Register if private key already exists
+- name: Register if config/private key already exists on target host
   stat:
-    path: "{{ private_key_file_path }}"
-  register: private_key_file_stat
-  delegate_to: localhost
+    path: "{{ wireguard_remote_directory }}/{{ wireguard_interface }}.conf"
+  register: config_file_stat
   tags:
     - wg-generate-keys
```
Comment (laurawadden) [inline tasks/main.yml]: I think you have to add the 'wg-config' tag here also.

File: tasks/main.yml [hunk 2]
```diff
@@ -28,106 +27,58 @@
...
-    - wg-generate-keys
-
-- name: Generate WireGuard public key
-  shell: "echo '{{ wg_private_key }}' | wg pubkey"
+- block:
+  - name: Generate WireGuard private key
+    shell: "wg genkey"
+    register: wg_private_key_result
+    tags:
+      - wg-generate-keys
```
Comment (laurawadden) [inline tasks/main.yml]: I would also add a `when: not config_file_stat.stat.exists` to this task as well as the one below it.

Comment (fbourqui) [reply tasks/main.yml]: No, it's a block, the wen apply to the full block:
```
- block:
  - name: Generate WireGuard private key
    shell: "wg genkey"
    register: wg_private_key_result
    tags:
      - wg-generate-keys

  - name: Set private key fact
    set_fact:
      private_key: "{{ wg_private_key_result.stdout }}"
    tags:
      - wg-generate-keys
  when: not config_file_stat.stat.exists
```

Comment (laurawadden) [reply tasks/main.yml]: Got it, thanks! I had accidentally indented my 'when'.

File: tasks/main.yml (hunk 1 above)
Comment (fbourqui) [reply tasks/main.yml]: No, we check if the config exist, to know if a key need to be generated, the config generation is done later with the generated key or the key existing in config file.

Comment (laurawadden) [reply tasks/main.yml]: But if you run `-t wg-config` alone without other tags, I thought that it needs to register whether or not the private key already exists.

Comment (fbourqui) [reply tasks/main.yml]: If you run -t wg-config, your assuming that the key exist already, as even is the config is not existing you would not generate a new one. So unless you merge tags "wg-config" and "wg-generate-keys", or put "wg-config" in all task with tag "wg-generate-keys

Comment (fbourqui) [reply tasks/main.yml]: OK, it make sense to add the tag, to not faill this block:

```
- block:
  - name: Read WireGuard config file
    slurp:
      src: "{{ wireguard_remote_directory }}/{{ wireguard_interface }}.conf"
    register: wg_config
    tags:
      - wg-config

  - name: Set private key fact
    set_fact:
      private_key: "{{ wg_config['content'] | b64decode | regex_findall('PrivateKey = (.*)') | first }}"
    tags:
      - wg-config
  when: config_file_stat.stat.exists
```

Comment (fbourqui) [reply tasks/main.yml]: Tad was added in latest commit.

File: tasks/main.yml [hunk 3]
```diff
@@ -28,106 +27,59 @@
...
+
+  - name: Set private key fact
+    set_fact:
+      private_key: "{{ wg_config['content'] | b64decode | regex_findall('PrivateKey = (.*)') | first }}"
+    tags:
+      - wg-config
+  when: config_file_stat.stat.exists
+
+- name: Derive WireGuard public key
+  shell: "echo '{{ private_key }}' | wg pubkey"
```
Comment (githubixx) [inline tasks/main.yml]:
```suggestion
  shell: "echo '{{ private_key }}' | wg pubkey" # noqa 306
```
ansible-lint complains that pipes should set the `pipefail` (bash) option which I would agree in general. But in that case it's not really needed as the last command `wg pubkey` will fail anyways if the `private_key` isn't valid. So by adding `noqa 306` ansible-lint will skip this check in that line.

**Comment by pierreozoux** [comment]
Thanks a lot @fbourqui :)

**Pull request merged**